third_party/move/move-vm/runtime/src/reentrancy_checker.rs

```diff
@@ -49,14 +50,17 @@ impl ReentrancyChecker {
     pub fn enter_function(
         &mut self,
         caller_module: Option<&ModuleId>,
-        callee_module: &ModuleId,
-        fun_name: &IdentStr,
+        callee: &LoadedFunction,
         call_type: CallType,
     ) -> PartialVMResult<()> {
-        if call_type == CallType::NativeDynamicDispatch {
-            // If we enter a native dispatch function, also enter module locking mode
+        if call_type == CallType::NativeDynamicDispatch
+            || callee.function.has_module_reentrancy_lock
```

Comment:
Would this not do the module locking when we have `#[module_lock]` for regular calls? Should we not do this only if: (a) it is a native dynamic dispatch, or (b) it is a closure dynamic dispatch and function has module reentrancy lock?

Otherwise we have some interesting behavior (that we may or may not want). If a top level function in the call graph has `#[module_lock]` and it calls a non-module locked function that has a closure dispatch with module re-entrancy, then the module lock property transfers to it.

Reply:
There might be strange behaviors, but it is also flexible because you can mark 'critical sections' within a callback graph. Let us see how the documentation comes out, its not necessarily finalized yet.